Update dnd examples to focus inserted items on drop

devongovett opened the pull request:
Behavior is:

* If dropping between two items, the dropped items are selected and the first one is focused.
* If dropping on an item, that item becomes focused and the focus ring is shown.
* When inserting between items but no items are added (e.g. they are reordered), the focus ring is shown only if the moved item is not selected.

majornista commented:
1. Should the dropped items be selected? I did this to match macOS finder behavior.

When dropping a single item in between items in the list, I don't think it should become selected unless it was selected before the drag and drop operation.

Since dragging multiple items requires the items to be selected, I think the behavior you have implemented, where items dropped between items in the list remain selected, and the first of the items has focus is appropriate.

When dropping one or more items on an item, like a folder, the drop target should display focus, but I don't think the drop target should become selected if it was not already selected.

2. Should the focus ring appear on mouse drops? My feeling was that if selection occurred this wasn't necessary, but if we don't select the dropped items then maybe it should to give some kind of feedback?

I think you're right about this, if we don't select the dropped items and there is no focus ring or focused state when you drop the items, it may be difficult for a mouse user to remember which items were dropped.

3. The current API requires the user do this manually. Is there a way we can make this automatic?

That would be ideal because, depending on the complexity of implementing this behavior manually, users might opt to omit it.

LFDanLu commented:
1. I think this is fine, matches with Windows behavior as well. As for the single item case that @majornista mentioned, I'm not entirely sure. Might be nice to auto select the dropped item for certain user flows e.g. could then do Shift+Arrow key commands immediately after to select a couple of other items in the droppable collection in preparation for a second drag/copy/cut operation.
2. It would definitely be nice to have some kind of indicator of where focus was currently located after a mouse drag operation. Windows has a light blue focus ring outline (in contrast to the filled blue box when selection is made) after a drag operation which reappears whenever you re-enter the file window via mouse or keyboard.
3.  No idea, I would need to play around with it to see if there is a way to do this automatically.

devongovett commented:
> When dropping a single item in between items in the list, I don't think it should become selected unless it was selected before the drag and drop operation.

I don't think we can know this in general, e.g. when dragging from Finder or somewhere outside the page. Feels to me like selecting is useful in most cases, especially when dropping multiple items since we can't focus multiple items at once.

> When dropping one or more items on an item, like a folder, the drop target should display focus, but I don't think the drop target should become selected if it was not already selected.

Yeah I agree. Updated to just show the focus ring instead when dropping on an item and no items are inserted.

> That would be ideal because, depending on the complexity of implementing this behavior manually, users might opt to omit it.

Yeah agreed. I added support for doing this by default, but there are some limitations. You must insert the items immediately on drop (within 50ms). When asynchronous behavior is involved things get very complicated very quickly. We can document how to do it manually for these cases, but I think this default should cover most use cases.

snowystinger commented on packages/@react-aria/dnd/src/useDroppableCollection.ts:
```diff
@@ -96,15 +103,90 @@ export function useDroppableCollection(props: DroppableCollectionOptions, state:
+    // and for React to re-render. If an insert occurs during this time, it will be selected/focused below.
+    // If items are not "immediately" inserted by the onDrop handler, the application will need to handle
+    // selecting and focusing those items themselves.
+    setTimeout(() => {
```
We need to save the timeout reference in case of unmount so we can cancel it

snowystinger commented on packages/@react-aria/dnd/src/useDroppableCollection.ts:
```diff
@@ -96,15 +103,90 @@ export function useDroppableCollection(props: DroppableCollectionOptions, state:
+      let {state} = localState;
+      if (state.selectionManager.focusedKey === focusedKey) {
+        if (target.type === 'item' && target.dropPosition === 'on') {
+          state.selectionManager.setFocusedKey(target.key);
```
do we need to make sure the target still exists?

snowystinger replied:
nevermind, that's covered by `state.collection.getItem(target.key) != null`

snowystinger commented on packages/@react-aria/dnd/stories/DroppableGrid.tsx:
```diff
@@ -135,6 +140,13 @@ function DroppableGrid(props) {
     })
   });
 
+  React.useImperativeHandle(ref, () => ({
```
why not destructure the import useImperativeHandle like we do for other React hooks?

devongovett replied:
🤷 it's just an example. Several other places in this file are like this.

snowystinger replied:
ah, yep, didn't notice it was in an example

snowystinger commented on packages/@react-aria/dnd/stories/Reorderable.tsx:
```diff
@@ -166,8 +142,17 @@ function ReorderableGrid(props) {
+        let keys = [];
+        for (let item of e.items) {
+          if (item.kind === 'text' && item.types.has(dragType)) {
+            let key = JSON.parse(await item.getText(dragType));
```
Interesting, do we have to wait for them sequentially? or could we throw them all into a Promise.all and run them in parallel?

```
     let promises = [];
     for (let item of e.items) {
          if (item.kind === 'text' && item.types.has(dragType)) {
            promises.push(item.getText(dragType));
     ...
     let keys = await Promise.all(promises).then(keys => keys.map(key => JSON.parse(key)));
     props.onMove(keys, e.target);
```

devongovett replied:
I don't think this really matters as reading the text is < 1ms. Just a story anyway.

snowystinger commented on packages/@react-aria/dnd/test/useDroppableCollection.test.js:
```diff
@@ -212,11 +212,15 @@ describe('useDroppableCollection', () => {
       });
 
-      await waitFor(() => expect(within(grid).getAllByRole('gridcell')).toHaveLength(4));
+      await waitFor(() => expect(within(grid).getAllByRole('gridcell')).toHaveLength(4), {interval: 10});
```
why 10? i don't see anything that advances timers by 10, also, how does this waitFor even work with faked timers like this? wouldn't it never check again after the first time (@t=0) no matter what interval you set?

devongovett replied:
I believe `waitFor` calls `advanceTimers` repeatedly on the given interval. I don't remember exactly, but the problem as I remember was that the default interval was jumping too far ahead, past our 50ms timer for the drop, so I lowered it.

snowystinger replied:
well, mind blown, thanks for the the explanation
